bmt/bmt.go

```diff
@@ -434,53 +465,78 @@ type SectionWriter interface {
 //   and keeps the resource which can be released calling Reset()
 type AsyncHasher struct {
-	*Hasher            // extends the Hasher
-	mtx     sync.Mutex // to lock the cursor access
-	double  bool       // whether to use double segments (call Hasher.writeSection)
-	secsize int        // size of base section (size of hash or double)
-	write   func(i int, section []byte, final bool)
+	*Hasher             // extends the Hasher
+	mtx      sync.Mutex // to lock the cursor access
+	double   bool       // whether to use double segments (call Hasher.writeSection)
+	secsize  int        // size of base section (size of hash or double)
+	seccount int        // base section count
+	write    func(i int, section []byte, final bool)
+	errFunc  func(error)
+	ctx      context.Context
+	all      bool // if all written in one go, temporary workaround
+}
+
+func (sw *AsyncHasher) raiseError(err string) {
+	if sw.errFunc != nil {
+		sw.errFunc(errors.New(err))
+	}
 }
 
-// methods needed to implement AsyncWriter
+// Reset implements file.SectionWriter
+func (sw *AsyncHasher) Reset() {
+	sw.all = false
+	sw.Hasher.Reset()
+}
 
-// SectionSize returns the size of async section unit to use
+// SectionSize implements file.SectionWriter
 func (sw *AsyncHasher) SectionSize() int {
 	return sw.secsize
 }
 
-// Write writes the i-th section of the BMT base
+// Branches implements file.SectionWriter
+func (sw *AsyncHasher) Branches() int {
+	return sw.seccount
+}
+
+// WriteSection writes the i-th section of the BMT base
 // this function can and is meant to be called concurrently
 // it sets max segment threadsafely
-func (sw *AsyncHasher) Write(i int, section []byte) {
+func (sw *AsyncHasher) WriteIndexed(i int, section []byte) {
 	sw.mtx.Lock()
 	defer sw.mtx.Unlock()
-	t := sw.getTree()
-	// cursor keeps track of the rightmost section written so far
+	t := sw.GetTree()
+	// cursor keeps track of the rightmost.GetSection() written so far
 	// if index is lower than cursor then just write non-final section as is
-	if i < t.cursor {
+	if i < sw.Hasher.GetCursor() {
 		// if index is not the rightmost, safe to write section
-		go sw.write(i, section, false)
+		go sw.WriteSection(i, section, sw.double, false)
 		return
 	}
-	// if there is a previous rightmost section safe to write section
-	if t.offset > 0 {
-		if i == t.cursor {
+	// if there is a previous rightmost.GetSection() safe to write section
+	if t.GetOffset() > 0 {
+		if i == sw.Hasher.GetCursor() {
 			// i==cursor implies cursor was set by Hash call so we can write section as final one
 			// since it can be shorter, first we copy it to the padded buffer
-			t.section = make([]byte, sw.secsize)
-			copy(t.section, section)
-			go sw.write(i, t.section, true)
+			//t.GetSection() = make([]byte, sw.secsize)
+			//copy(t.GetSection(), section)
+			// TODO: Consider whether the section here needs to be copied, maybe we can enforce not change the original slice
+			copySection := make([]byte, sw.secsize)
```

Comment:
why the copying not part of SetSection then?

Reply:
There's no member in `tree` that remembers the section size, so either we must add a member or we must pass it with the function. The latter seems clumsy.

In general I think it's a good idea to introduce as few side-effects in the lower level components as possible; the tree component _could_ be used without any copying, after all.